packages/deployments/contracts/contracts/facets/StableSwapFacet.sol

```diff
@@ -0,0 +1,458 @@
+
+    uint256[] memory precisionMultipliers = new uint256[](decimals.length);
+
+    for (uint8 i = 0; i < _pooledTokens.length; i++) {
+      if (i > 0) {
+        // Check if index is already used. Check if 0th element is a duplicate.
+        if (s.tokenIndexes[_canonicalId][address(_pooledTokens[i])] != 0 || _pooledTokens[0] == _pooledTokens[i])
+          revert StableSwapFacet__initializeSwap_duplicateTokens();
+      }
+      if (address(_pooledTokens[i]) == address(0)) revert StableSwapFacet__initializeSwap_zeroTokenAddress();
+
+      if (decimals[i] > SwapUtils.POOL_PRECISION_DECIMALS)
+        revert StableSwapFacet__initializeSwap_tokenDecimalsExceedMax();
+
+      precisionMultipliers[i] = 10**uint256(SwapUtils.POOL_PRECISION_DECIMALS - decimals[i]);
+      s.tokenIndexes[_canonicalId][address(_pooledTokens[i])] = i;
+    }
+
+    // Check _a, _fee, _adminFee, _withdrawFee parameters
+    if (_a >= AmplificationUtils.MAX_A) revert StableSwapFacet__initializeSwap_aExceedMax();
+    if (_fee >= SwapUtils.MAX_SWAP_FEE) revert StableSwapFacet__initializeSwap_feeExceedMax();
+    if (_adminFee >= SwapUtils.MAX_ADMIN_FEE) revert StableSwapFacet__initializeSwap_adminFeeExceedMax();
+
+    // Initialize a LPToken contract
+    LPToken lpToken = LPToken(Clones.clone(lpTokenTargetAddress));
+    if (!lpToken.initialize(lpTokenName, lpTokenSymbol))
+      revert StableSwapFacet__initializeSwap_failedInitLpTokenClone();
+
+    // Initialize swapStorage struct
+    s.swapStorages[_canonicalId] = SwapUtils.Swap({
+      initialA: _a * AmplificationUtils.A_PRECISION,
+      futureA: _a * AmplificationUtils.A_PRECISION,
+      swapFee: _fee,
+      adminFee: _adminFee,
+      lpToken: lpToken,
+      pooledTokens: _pooledTokens,
+      tokenPrecisionMultipliers: precisionMultipliers,
+      balances: new uint256[](_pooledTokens.length),
+      initialATime: 0,
+      futureATime: 0
+    });
+  }
+
+  /**
+   * @notice Withdraw all admin fees to the contract owner
+   * @param canonicalId the canonical token id
+   */
+  function withdrawAdminFees(bytes32 canonicalId) external onlyOwner {
+    s.swapStorages[canonicalId].withdrawAdminFees(msg.sender);
+  }
+
+  /**
+   * @notice Update the admin fee. Admin fee takes portion of the swap fee.
+   * @param canonicalId the canonical token id
+   * @param newAdminFee new admin fee to be applied on future transactions
+   */
+  function setAdminFee(bytes32 canonicalId, uint256 newAdminFee) external onlyOwner {
+    s.swapStorages[canonicalId].setAdminFee(newAdminFee);
+  }
+
+  /**
+   * @notice Update the swap fee to be applied on swaps
+   * @param canonicalId the canonical token id
+   * @param newSwapFee new swap fee to be applied on future transactions
+   */
+  function setSwapFee(bytes32 canonicalId, uint256 newSwapFee) external onlyOwner {
+    s.swapStorages[canonicalId].setSwapFee(newSwapFee);
+  }
```

Comment:
I'm assuming that these fee setter methods will be delegated to be only governance in the future

fee changes should be subject to a commitment period as well (like submit the change, and then commit it 72 hours later)

just something to keep in mind